File: dailyHOTEL/app/src/main/java/com/twoheart/dailyhotel/screen/information/coupon/SelectCouponAdapter.java
package com.twoheart.dailyhotel.screen.information.coupon;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.TextView;

import com.crashlytics.android.Crashlytics;
import com.twoheart.dailyhotel.R;
import com.twoheart.dailyhotel.model.Coupon;
import com.twoheart.dailyhotel.util.Constants;
import com.twoheart.dailyhotel.util.DailyCalendar;
import com.twoheart.dailyhotel.util.ExLog;
import com.twoheart.dailyhotel.util.Util;
import com.twoheart.dailyhotel.widget.DailyImageView;
import com.twoheart.dailyhotel.widget.DailyTextView;

import java.text.ParseException;
import java.util.List;

/**
 * Created by android_sam on 2016. 5. 26..
 */
public class SelectCouponAdapter extends ArrayAdapter<Coupon>
{
    private static final int RESOURCE_ID = R.layout.list_row_select_coupon;

    private Context mContext;
    private List<Coupon> mList;
    private OnCouponItemListener mListener;
    private int mSelectPosition = -1;
    private boolean mIsSelectedMode;

    public interface OnCouponItemListener
    {
        void onDownloadClick(int position);

        void updatePositiveButton();
    }

    public SelectCouponAdapter(Context context, List<Coupon> list, OnCouponItemListener listener)
    {
        super(context, RESOURCE_ID, list);

        mContext = context;
        mList = list;
        mListener = listener;
        setSelectedMode(true);
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent)
    {
        convertView = newView(convertView);

        if (convertView != null)
        {
            bindView(position, convertView);
        }

        return convertView;
    }


    private View newView(View convertView)
    {
        if (convertView != null)
        {
            return convertView;
        }

        View view = LayoutInflater.from(mContext).inflate(RESOURCE_ID, null);

        SelectViewHolder holder = new SelectViewHolder();

        holder.couponLayout = view;
        holder.priceTextView = (DailyTextView) view.findViewById(R.id.priceTextView);
        holder.titleTextView = (TextView) view.findViewById(R.id.titleTextView);
        holder.minPriceTextView = (TextView) view.findViewById(R.id.minPriceTextView);
        holder.expireTextView = (TextView) view.findViewById(R.id.expireTextView);
        holder.hotelIconView = (DailyImageView) view.findViewById(R.id.hotelIconView);
        holder.gourmetIconView = (DailyImageView) view.findViewById(R.id.gourmetIconView);
        holder.downloadCouponView = view.findViewById(R.id.downloadCouponView);
        holder.underLineView = view.findViewById(R.id.underLineView);
        view.setTag(holder);

        return view;
    }

    private void bindView(final int position, View convertView)
    {
        final Object tag = convertView.getTag();
        if (tag == null || !(tag instanceof SelectViewHolder))
        {
            return;
        }

        SelectViewHolder holder = (SelectViewHolder) convertView.getTag();

        Coupon coupon = getItem(position);

        String strAmount = Util.getPriceFormat(mContext, coupon.amount, false);
        holder.priceTextView.setText(strAmount);

        holder.titleTextView.setText(coupon.title);

        if (coupon.amountMinimum > 0)
        {
            String strAmountMinimum = mContext.getResources().getString( //
                R.string.coupon_min_price_text, //
                Util.getPriceFormat(mContext, coupon.amountMinimum, false));

            holder.minPriceTextView.setText(strAmountMinimum);
            holder.minPriceTextView.setVisibility(View.VISIBLE);
        } else
        {
            holder.minPriceTextView.setText("");
            holder.minPriceTextView.setVisibility(View.GONE);
        }

        try
        {
            String expireText = DailyCalendar.convertDateFormatString(coupon.validTo, DailyCalendar.ISO_8601_FORMAT, "yyyy.MM.dd");
            expireText = String.format("~%s", expireText);
            holder.expireTextView.setText(expireText);
            holder.expireTextView.setVisibility(View.VISIBLE);
        } catch (ParseException e)
        {
            if (Constants.DEBUG == false)
            {
                Crashlytics.log("Select Coupon::coupon.vaildTo: " + (coupon != null ? coupon.validTo : ""));
            }
            ExLog.d(e.getMessage());
            holder.expireTextView.setVisibility(View.GONE);
        }

        // 스테이, 고메 쿠폰인지
        if (coupon.availableInHotel == true)
        {
            holder.hotelIconView.setVectorImageResource(R.drawable.ic_badge_hotel_on);
        } else
        {
            holder.hotelIconView.setVectorImageResource(R.drawable.ic_badge_hotel_off);
        }

        if (coupon.availableInGourmet == true)
        {
            holder.gourmetIconView.setVectorImageResource(R.drawable.ic_badge_gourmet_on);
        } else
        {
            holder.gourmetIconView.setVectorImageResource(R.drawable.ic_badge_gourmet_off);
        }

        setDownLoadLayout(holder, coupon.isDownloaded);

        if (mIsSelectedMode == true && coupon.isDownloaded == true)
        {
            setSelectedLayout(holder, mSelectPosition == position);

            holder.couponLayout.setOnClickListener(new View.OnClickListener()
            {
                @Override
                public void onClick(View v)
                {
                    Coupon coupon = getItem(position);
                    if (coupon.isDownloaded == true)
                    {
                        mSelectPosition = position;
                    } else
                    {
                        mSelectPosition = -1;
                        mListener.onDownloadClick(position);
                    }

                    notifyDataSetChanged();
                    mListener.updatePositiveButton();
                }
            });
        } else
        {
            holder.couponLayout.setOnClickListener(new View.OnClickListener()
            {
                @Override
                public void onClick(View v)
                {
                    Coupon coupon = getItem(position);
                    if (coupon.isDownloaded == false)
                    {
                        mListener.onDownloadClick(position);
                    }
                }
            });
        }
    }

    public void setData(List<Coupon> list)
    {
        mList = list;
        clear();
        addAll(list);
    }

    public void setSelectedMode(boolean selected)
    {
        mIsSelectedMode = selected;
    }

    public Coupon getCoupon(String userCouponCode)
    {
        if (mList == null)
        {
            return null;
        }

        for (Coupon coupon : mList)
        {
            if (coupon.userCouponCode.equalsIgnoreCase(userCouponCode) == true)
            {
                return coupon;
            }
        }

        return null;
    }

    public int getSelectPosition()
    {
        return mSelectPosition;
    }

    private void setSelectedLayout(SelectViewHolder holder, boolean isSelected)
    {
        if (isSelected == true)
        {
            holder.couponLayout.setBackgroundColor(mContext.getResources().getColor(R.color.default_background_cfafafb));

            holder.priceTextView.setTextColor(mContext.getResources().getColor(R.color.default_text_c900034));
            holder.priceTextView.setCompoundDrawablesWithIntrinsicBounds(0, 0, R.drawable.ic_check_s, 0);

            holder.titleTextView.setTextColor(mContext.getResources().getColor(R.color.default_text_c900034));
        } else
        {
            holder.couponLayout.setBackgroundResource(0);

            holder.priceTextView.setTextColor(mContext.getResources().getColor(R.color.default_text_c323232));
            holder.priceTextView.setCompoundDrawablesWithIntrinsicBounds(0, 0, 0, 0);

            holder.titleTextView.setTextColor(mContext.getResources().getColor(R.color.default_text_c323232));
        }
    }

    private void setDownLoadLayout(SelectViewHolder holder, boolean isDownload)
    {
        if (isDownload == true)
        {
            holder.priceTextView.setTextColor(mContext.getResources().getColor(R.color.default_text_c323232));
            holder.titleTextView.setTextColor(mContext.getResources().getColor(R.color.default_text_c323232));
            holder.minPriceTextView.setTextColor(mContext.getResources().getColor(R.color.default_text_c929292));
            holder.expireTextView.setTextColor(mContext.getResources().getColor(R.color.default_text_c929292));

            holder.hotelIconView.setAlpha(1.0f);
            holder.gourmetIconView.setAlpha(1.0f);
            holder.downloadCouponView.setVisibility(View.GONE);

        } else
        {
            int color = mContext.getResources().getColor(R.color.default_text_cc5c5c5);

            holder.priceTextView.setTextColor(color);
            holder.titleTextView.setTextColor(color);
            holder.minPriceTextView.setTextColor(color);
            holder.expireTextView.setTextColor(color);

            holder.hotelIconView.setAlpha(0.5f);
            holder.gourmetIconView.setAlpha(0.5f);
            holder.downloadCouponView.setVisibility(View.VISIBLE);
        }
    }

    protected class SelectViewHolder
    {
        View couponLayout;
        DailyTextView priceTextView;
        TextView titleTextView;
        TextView minPriceTextView;
        TextView expireTextView;
        DailyImageView hotelIconView;
        DailyImageView gourmetIconView;
        View downloadCouponView;
        View underLineView;
    }
}

File: dailyHOTEL/app/src/main/java/com/daily/dailyhotel/screen/mydaily/coupon/dialog/SelectCouponAdapter.kt
package com.daily.dailyhotel.screen.mydaily.coupon.dialog

import android.content.Context
import android.databinding.DataBindingUtil
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.crashlytics.android.Crashlytics
import com.daily.base.util.DailyTextUtils
import com.daily.base.util.ExLog
import com.daily.dailyhotel.entity.Coupon
import com.daily.dailyhotel.util.isTextEmpty
import com.twoheart.dailyhotel.R
import com.twoheart.dailyhotel.databinding.LayoutSelectCouponDataBinding
import com.twoheart.dailyhotel.util.DailyCalendar
import java.text.ParseException
import java.util.*

class SelectCouponAdapter(val context: Context, val list: MutableList<Coupon>, val listener: OnCouponItemLister) : RecyclerView.Adapter<SelectCouponAdapter.CouponViewHolder>() {
    var selectedPosition = -1
    var selectedMode = true

    interface OnCouponItemLister {
        fun onDownloadClick(position: Int)

        fun updatePositiveButton()
    }

    override fun getItemCount(): Int {
        return list.size
    }

    fun clear() {
        list.clear()
    }

    fun addAll(list: MutableList<Coupon>) {
        this.list += list
    }

    fun setAll(list: MutableList<Coupon>) {
        clear()
        addAll(list)
    }

    fun getItem(position: Int): Coupon? {
        return (position < 0 || list.size <= position).let { null } ?: list[position]
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CouponViewHolder {
        val viewDataBinding = DataBindingUtil.inflate<LayoutSelectCouponDataBinding>(LayoutInflater.from(context), R.layout.layout_select_coupon_data, parent, false)
        return CouponViewHolder(viewDataBinding)
    }

    override fun onBindViewHolder(holder: CouponViewHolder, position: Int) {
        val coupon = getItem(position) ?: return

        holder.dataBinding.run {
            val amount = DailyTextUtils.getPriceFormat(context, coupon.amount, false)
            couponPriceTextView.text = amount
            couponNameTextView.text = coupon.title

            descriptionTextView.run {
                when (coupon.amountMinimum) {
                    0 -> {
                        visibility = View.GONE
                    }

                    else -> {
                        text = context.resources.getString(R.string.coupon_min_price_text, DailyTextUtils.getPriceFormat(context, coupon.amountMinimum, false))
                        visibility = View.VISIBLE
                    }
                }
            }

            expireTextView.run {
                try {
                    var expireText = DailyCalendar.convertDateFormatString(coupon.validTo, DailyCalendar.ISO_8601_FORMAT, "yyyy.MM.dd")
                    expireText = String.format(Locale.KOREA, "- %s", expireText)
                    text = expireText
                    visibility = View.VISIBLE
                } catch (e: ParseException) {
                    Crashlytics.log("Select Coupon::coupon.validTo: " + coupon.validTo)
                    ExLog.d(e.message)
                    visibility = View.GONE
                }
            }

            // stay, gourmet, StayOutbound Type
            useableStayTextView.visibility = if (coupon.availableInStay) View.VISIBLE else View.GONE
            useableStayOutboundTextView.visibility = if (coupon.availableInOutboundHotel) View.VISIBLE else View.GONE
            useableGourmetTextView.visibility = if (coupon.availableInGourmet) View.VISIBLE else View.GONE

            setDownLoadLayout(this, Coupon.Type.REWARD == coupon.type, coupon.isDownloaded)

            if (selectedMode && coupon.isDownloaded) {
                setSelectedLayout(this, Coupon.Type.REWARD == coupon.type, selectedPosition == position)

                root.setOnClickListener {
                    val coupon = getItem(position)

                    if (coupon != null && coupon.isDownloaded) {
                        selectedPosition = position
                    } else {
                        selectedPosition = -1
                        listener.onDownloadClick(position)
                    }

                    notifyDataSetChanged()
                    listener.updatePositiveButton()
                }
            } else {
                root.setOnClickListener {
                    val coupon = getItem(position)

                    coupon?.let {
                        if (!it.isDownloaded) {
                            listener.onDownloadClick(position)
                        }
                    }
                }
            }
        }
    }

    fun getCoupon(couponCode: String?): Coupon? {
        if (couponCode.isTextEmpty()) {
            return null
        }

        if (list.isEmpty()) {
            return null
        }

        for (coupon in list) {
            if (couponCode.equals(coupon.couponCode, true)) {
                return coupon
            }
        }

        return null
    }

    private fun setSelectedLayout(dataBinding: LayoutSelectCouponDataBinding, rewardCoupon: Boolean, selected: Boolean) {
        dataBinding.run {
            when (selected) {
                true -> {
                    root.setBackgroundColor(context.resources.getColor(R.color.default_background_cfafafb))

                    couponPriceTextView.run {
                        setTextColor(context.resources.getColor(R.color.default_text_ceb2135))
                        setCompoundDrawablesWithIntrinsicBounds(if (rewardCoupon) R.drawable.vector_r_ic_s_17 else 0, 0, R.drawable.ic_check_s, 0)
                        isSelected = true
                    }

                    couponNameTextView.setTextColor(context.resources.getColor(R.color.default_text_ceb2135))
                }

                else -> {
                    root.setBackgroundResource(R.color.white)

                    couponPriceTextView.run {
                        setTextColor(context.resources.getColor(R.color.black))
                        setCompoundDrawablesWithIntrinsicBounds(if (rewardCoupon) R.drawable.vector_r_ic_s_17 else 0, 0, 0, 0)
                        isSelected = false
                    }

                    couponNameTextView.setTextColor(context.resources.getColor(R.color.default_text_c4d4d4d))
                }
            }
        }
    }

    private fun setDownLoadLayout(dataBinding: LayoutSelectCouponDataBinding, rewardCoupon: Boolean, download: Boolean) {
        dataBinding.run {
            couponPriceTextView.setCompoundDrawablesWithIntrinsicBounds(if (rewardCoupon) R.drawable.vector_r_ic_s_17 else 0, 0, 0, 0)
            root.setBackgroundResource(R.color.white)

            when (download) {
                true -> {
                    couponPriceTextView.setTextColor(context.resources.getColor(R.color.black))
                    couponNameTextView.setTextColor(context.resources.getColor(R.color.black))
                    descriptionTextView.setTextColor(context.resources.getColor(R.color.default_text_c929292))
                    expireTextView.setTextColor(context.resources.getColor(R.color.default_text_c929292))

                    useableStayTextView.alpha = 1.0f
                    useableStayOutboundTextView.alpha = 1.0f
                    useableGourmetTextView.alpha = 1.0f
                    downloadCouponLayout.visibility = View.GONE
                }

                else -> {
                    val color = context.resources.getColor(R.color.default_text_cc5c5c5)

                    couponPriceTextView.setTextColor(color)
                    couponNameTextView.setTextColor(color)
                    descriptionTextView.setTextColor(color)
                    expireTextView.setTextColor(color)

                    useableStayTextView.alpha = 0.5f
                    useableStayOutboundTextView.alpha = 0.5f
                    useableGourmetTextView.alpha = 0.5f
                    downloadCouponLayout.visibility = View.VISIBLE
                }
            }
        }
    }

    inner class CouponViewHolder(val dataBinding: LayoutSelectCouponDataBinding) : RecyclerView.ViewHolder(dataBinding.root)
}